lib/s3routes/routesUtils.js

```diff
@@ -242,54 +247,97 @@ function okContentHeadersResponse(overrideParams, resHeaders,
-                logger.error('failed to get object from source', {
-                    error: err,
-                    method: 'retrieveData',
-                    backend: 'Azure',
+    // response is of type http.ServerResponse
+    let responseDestroyed = false;
+    let currentStream = null; // reference to the stream we are reading from
+    const _destroyResponse = () => {
+        // destroys the socket if available
+        response.destroy();
+        responseDestroyed = true;
+    };
+    // the S3-client might close the connection while we are processing it
+    response.once('close', () => {
+        log.debug('received close event before response end');
+        responseDestroyed = true;
+        if (currentStream) {
+            currentStream.destroy();
+        }
+    });
+
+    return eachSeries(locations,
+        (current, next) => retrieveDataFn(current, response, log,
+            (err, readable) => {
+                // NB: readable is of IncomingMessage type
+                if (err) {
+                    log.error('failed to get object', {
+                        error: err,
+                        method: 'retrieveData',
+                    });
+                    _destroyResponse();
+                    return next(err);
+                }
+                // response.isclosed is set by the S3 server. Might happen if
+                // the S3-client closes the connection before the first request
+                // to the backend is started.
+                if (responseDestroyed || response.isclosed) {
+                    log.debug(
+                        'response destroyed before readable could stream');
+                    readable.destroy();
+                    return next(responseErr);
+                }
+                // readable stream successfully consumed
+                readable.on('end', () => {
+                    currentStream = null;
+                    log.debug('readable stream end reached');
+                    return next();
                 });
-                return errorHandlerFn(err);
-            }
-            return undefined;
-        });
-    }
-    return dataRetrievalFn(current, response, logger,
-        (err, readable) => {
-            if (err) {
-                logger.error('failed to get object', {
-                    error: err,
-                    method: 'retrieveData',
+                // errors on server side with readable stream
+                readable.on('error', err => {
+                    log.error('error piping data from source');
+                    _destroyResponse();
+                    return next(err);
                 });
-                return errorHandlerFn(err);
+                currentStream = readable;
+                return readable.pipe(response, { end: false });
+            }), err => {
+                currentStream = null;
+                if (err) {
+                    log.debug('abort response due to client error', {
```

Comment:
minor, especially since it's a debug log, but you can also have an error due to the readable object here.